### DISCUSSION TITLE
RFC 1130 - additions to collection functions in Array.Parallel

### DISCUSSION THREAD

**Discussion by T-Gro** [fsharp/fslang-design#719] [Language and core library RFC discussions]
This is a discussion for RFC 1130 building on [Make FSharp.Core collection functions for Array.Parallel more regular](https://github.com/fsharp/fslang-suggestions/issues/187).

The fslang suggestions uses the phrase " lots of **standard** functions are missing" and I would like to use this discussion to evaluate the scope of the addition, i.e. to see what is generally understoof by "standard" and where people see the biggest benefit of Array.Parallel on top of the proposed functions.

All the functions mentioned in the suggestion make sense and I marked them as "ADD"  (to be added by this extension).
I excluded all functions about appending/removing/updating data, and also all functions that are already O(1) even without using parallelism.

For the rest, I would be tempted to not implement them unless strongly voted for here or elsewhere.
Thinking about it, the fslang-design PR itself might be a better place to raise points to individual comments, because of the "review line-by-line" nature which a PR enables.

**Comment by voronoipotato** [comment]
## Add
- Exists
- Filter
- ForAll
- MaxBy
- MinBy
- Reduce
- TryFind
- TryFindIndex
- TryPick

Have we skimmed through other Parallel implementations for what they have, make sure we're not missing something that would be helpful http://fsprojects.github.io/FSharp.Collections.ParallelSeq/ comes to mind. Otherwise I think this looks fine.

**Reply by voronoipotato** [reply]
## Things I would also want
- Array.Parallel.Sum
    - it's commutative my intuition is this can be made faster than `reduce(+)` 
- Array.Parallel.GroupBy 
    - should be faster in parallel, pretty useful
- Array.Parallel.Sort 
    - where sort is parallel, ought to be faster, very useful
- Array.Parallel.Zip
    - if we don't have map2 we should at least have zip, seems like it may be faster in parallel 

### Out of scope ideas, may need new proposal mostly putting here so I don't later forget. 
- Not sure if these are needed for Array.Parallel or just LINQ but I thought I'd mention it
    - Array.Parallel.WithCancellation
    - Array.Parallel.WithDegreeOfParallelism
    - Array.Parallel.WithExecutionMode
    - Array.Parallel.WithParallelMergeOptions
- Array.Parallel.Cache seems useful, but that's probably covered by AVal now.

**Comment by T-Gro** [comment]
Thanks, the functional additions make sense.

The "With" ones though, that would imply keeping state around and I do not see it as an F# idiomatic way.
LINQ has those , and LINQ can be used/wrapped by F# as well -> this PR does not try to neither replace LINQ nor wrap it, it should be a standalone implementation for Array.Parallel in F#.

**Reply by voronoipotato** [reply]
sounds good to me

**Reply by T-Gro** [reply]
( I have incorporated the 4 additions to the RFC + expanded the "sort" to the full family of sort - sort,sortBy,sortDescending,sortByDescending)
